app/src/main/java/org/olpc/xorobothash/MainActivity.java

```diff
@@ -0,0 +1,157 @@
+
+    private void showDisconnected() {
+        //Hide the robot text
+        TextView robotText = (TextView) findViewById(R.id.robot_text_message);
+        robotText.setVisibility(View.GONE);
+        //Hide the robot image
+        ImageView robotImage = (ImageView) findViewById(R.id.robot_hash_image);
+        robotImage.setVisibility(View.GONE);
+        //Show the disconnected error message
+        TextView errorMessage = (TextView) findViewById(R.id.error_message);
+        errorMessage.setText(getString(R.string.no_network_connection));
+    }
+
+    private String robotImagePath() {
+        return this.getFilesDir() + File.pathSeparator + "this-xos-robot.png";
+    }
+
+    private String uniqueRoboHashURL() {
+        //TODO test on a real android device
+        String serialNumber = android.os.Build.SERIAL;  // if SERIAL doesn't work, try Settings.Secure.ANDROID_ID; ?
+        return "http://robohash.org/" + serialNumber + ".png?size=500x500";
+    }
+
+    private boolean checkNetworkConnectivity() {
+        boolean connected;
+        ConnectivityManager connMgr = (ConnectivityManager)
+                getSystemService(Context.CONNECTIVITY_SERVICE);
+        NetworkInfo networkInfo = connMgr.getActiveNetworkInfo();
+        if (networkInfo != null && networkInfo.isConnected()) {
+            Log.i(TAG, "Network is connected.");
+            connected = true;
+        } else {
+            Log.e(TAG, "Network is NOT connected.");
+            connected = false;
+        }
+        return connected;
+    }
+
+    private class DownloadRobotImage extends AsyncTask<String, Void, Boolean> {
+
+        @Override
+        protected void onPreExecute() {
+            // TODO Display a spinner or loading message
+        }
+
+        @Override
+        protected Boolean doInBackground(String... params) {
+            Log.i(TAG, "In DownloadRobotImage() doInBackground()");
+            HttpURLConnection connection = null;
+            Boolean downloadSuccess = true;
+            try {
+                URL url = new URL(params[0]);
+                connection = (HttpURLConnection) url.openConnection();
+
+                InputStream in = new BufferedInputStream(connection.getInputStream());
+                OutputStream out = new FileOutputStream(params[1]);
+
+                int data = in.read();
+                while (data != -1) {
+                    out.write(data);
+                    data = in.read();
+                }
+            } catch (IOException e) {
+                Log.e(TAG + " TASK", "error retrieving robohash.org image", e);
+                downloadSuccess = false;
+            } finally {
+                if (connection != null) {
+                    connection.disconnect();
+                }
+            }
+            return downloadSuccess;
+        }
+
+        protected void onPostExecute(Boolean downloadSuccess) {
+            Log.i(TAG, "In onPostExecute()");
+            updateView(downloadSuccess);
```

Comment:
I think I have a problem here.
Assume the `doInBackground()` is unsuccessful and returns `false`, so `updateView(false)` is called. 
`updateView(false)` could still determine a potentially incomplete `file.exists()`, and proceed to call `showRobot()`. The error message from `showDisconnected()` is never displayed, and we are forever stuck with a potentially malformed robot.